docs/decisions/0045-updated-vector-store-design.md

```diff
@@ -0,0 +1,749 @@
+
+
+### Memory Store Today
+```cs
+interface IMemoryStore
+{
+    // Collection / Index Management
+    Task CreateCollectionAsync(string collectionName, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<string> GetCollectionsAsync(CancellationToken cancellationToken = default);
+    Task<bool> DoesCollectionExistAsync(string collectionName, CancellationToken cancellationToken = default);
+    Task DeleteCollectionAsync(string collectionName, CancellationToken cancellationToken = default);
+
+    // Data Storage and Retrieval
+    Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records, CancellationToken cancellationToken = default);
+    Task<MemoryRecord?> GetAsync(string collectionName, string key, bool withEmbedding = false, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<MemoryRecord> GetBatchAsync(string collectionName, IEnumerable<string> keys, bool withVectors = false, CancellationToken cancellationToken = default);
+    Task RemoveAsync(string collectionName, string key, CancellationToken cancellationToken = default);
+    Task RemoveBatchAsync(string collectionName, IEnumerable<string> keys, CancellationToken cancellationToken = default);
+
+    // Vector Search
+    IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(
+        string collectionName,
+        ReadOnlyMemory<float> embedding,
+        int limit,
+        double minRelevanceScore = 0.0,
+        bool withVectors = false,
+        CancellationToken cancellationToken = default);
+
+    Task<(MemoryRecord, double)?> GetNearestMatchAsync(
+        string collectionName,
+        ReadOnlyMemory<float> embedding,
+        double minRelevanceScore = 0.0,
+        bool withEmbedding = false,
+        CancellationToken cancellationToken = default);
+}
+```
+
+### Actions
+
+1. The `IMemoryStore` should be split into four different interfaces, one for each responsibility.
+2. The **Data Storage and Retrieval** and **Vector Search** areas should allow typed access to data and support any schema that is currently available in the customer's data store.
+3. The collection / index create functionality should allow developers to create their own implementations and support creating first party collections for built in functionality. Each implementation would be for a specific schema and data store type.
+4. The collection / index list/exists/delete functionality should allow management of any collection regardless of schema. There should be one implementation for each data store type.
+5. Remove opinionated behaviors from connectors. The opinionated behavior limits the ability of these connectors to be used with pre-existing vector databases. As far as possible these behaviors should be moved into decorators or be injectable.  Examples of opinionated behaviors:
```

Comment:
One reason for past opionated behavior was the ability to switch from one DB to another, without changing code. 
For instance, Qdrant supports only Guid/Numeric IDs, while Azure AI Search supports strings. When working locally as a developer I need to use Qdrant to save money, and when I deploy my code I need to use Azure AI Search to scale and all the other benefits. 
As a developer I think the ideal solution allows to switch from Qdrant to Azure AI Search without changing code (e.g. GUID vs strings), for example just changing the dependency injection settings.

Reply:
On a related note, there should be a consistent set of exceptions thrown by connectors, so that try-catch blocks are agnostic about the underlying connector used.

Reply:
> especially if they have already created their own database and want to read/write to it using our code

I think we're on the same page here. Bring your own datasource is a top priority. My preference on this matter is to address this with configuration ((which field is what, aka mappers). I've seen attempts at using custom classes to map custom schemas, and the code get very complex very quickly, e.g. due to an excessive use of generics that are coupled to specific storage types. In my experience, relying on configuration makes it very easy to switch storage without changing code, and allows to use the same approach across different programming languages.